c++/src/kj/async-io.h

```diff
@@ -106,6 +106,17 @@ class AsyncOutputStream {
   // output stream. If it finds one, it performs the pump. Otherwise, it returns null.
   //
   // The default implementation always returns null.
+
+  virtual Promise<void> whenWriteDisconnected() = 0;
+  // Returns a promise that resolves when the stream has become disconnected such that new write()s
+  // will fail with a DISCONNECTED exception. This is particularly useful, for example, to cancel
+  // work early when it is detected that no one will receive the result.
+  //
+  // Note that not all streams are able to detect this condition without actually performing a
+  // write(); such stream implementations may return a promise that never resolves.
```

Comment:
In the comment I was more thinking about it being unimplemented on Windows IOCP. (And it looks like it may not work on OSX or Cygwin either...)

TCP streams actually should usually be able to detect the condition. If a user clicks "stop" or navigates away in the browser, that should send an appropriate reset to the server. If a client just disappears (e.g. power/network outage), the problem can still be detected after a timeout if TCP keepalive has been enabled on the socket.

I'll add comments once I finish debugging OSX...